From 009ecc71a0b397faed8fafb3c4d62e96d0d86a8a Mon Sep 17 00:00:00 2001
From: Keshav Padram Amburay <address@hidden>
Modified for Manjaro by: Philip Müller <address@hidden>
Date: Wed, 6 Nov 2013 21:49:40 +0530

Additional modification for GRUB_ROOT_FS_MODE by: 10101000 <address@hidden>
Date: Fri, 11 Dec 2020 14:03:57 -0700

Subject: [PATCH] Add Arch Linux specific grub-mkconfig fixes

Patch modified based on ideas from Felix aka fstirlitz,
given at https://bugs.archlinux.org/task/37904?getfile=11257
---

diff -Npur a/util/grub-mkconfig.in b/util/grub-mkconfig.in
--- a/util/grub-mkconfig.in	2026-10-06 14:02:08.637184625 -0600
+++ b/util/grub-mkconfig.in	2026-10-06 14:06:08.579177305 -0600
@@ -285,7 +285,8 @@ export GRUB_DEFAULT \
   GRUB_SIMPLE_ACCESS_LEVEL \
   GRUB_ADVANCED_ACCESS_LEVEL \
   GRUB_RECOVERY_ACCESS_LEVEL \
-  GRUB_BOOTNEXT_OMIT_INACTIVE
+  GRUB_BOOTNEXT_OMIT_INACTIVE \
+  GRUB_ROOT_FS_RO
 
 if test "x${grub_cfg}" != "x"; then
   rm -f "${grub_cfg}.new"
diff -Npur a/util/grub.d/10_linux.in b/util/grub.d/10_linux.in
--- a/util/grub.d/10_linux.in	2026-10-06 13:23:35.699732104 -0600
+++ b/util/grub.d/10_linux.in	2026-10-06 13:39:18.854635633 -0600
@@ -100,11 +100,15 @@ linux_entry ()
   if [ x$type != xsimple ] ; then
       case $type in
 	  recovery)
-	      title="$(gettext_printf "%s, with Linux %s (recovery mode)" "${os}" "${version}")"
+	      title="$(gettext_printf "%s (Kernel: %s - recovery mode)" "${os}" "${version}")"
+	      allowed_user_list="$(grub_get_auth_user_list $GRUB_RECOVERY_ACCESS_LEVEL)"
+	      ;;
+	  fallback)
+	      title="$(gettext_printf "%s (Kernel: %s - fallback initramfs)" "${os}" "${version}")"
 	      allowed_user_list="$(grub_get_auth_user_list $GRUB_RECOVERY_ACCESS_LEVEL)"
 	      ;;
 	  *)
-	      title="$(gettext_printf "%s, with Linux %s" "${os}" "${version}")"
+	      title="$(gettext_printf "%s (Kernel: %s)" "${os}" "${version}")"
 	      allowed_user_list="$(grub_get_auth_user_list $GRUB_ADVANCED_ACCESS_LEVEL)"
 	      ;;
       esac
@@ -106,7 +108,7 @@ linux_entry ()
   else
       echo "menuentry '$(echo "$os" | grub_quote)' ${CLASS} \$menuentry_id_option 'gnulinux-simple-$boot_device_id' {" | sed "s/^/$submenu_indentation/"
   fi      
-  if [ x$type != xrecovery ] ; then
+  if [ x$type != xrecovery ] && [ x$type != xfallback ] ; then
       save_default_entry | grub_add_tab
   fi
 
@@ -138,10 +140,17 @@ linux_entry ()
     fi
     printf '%s\n' "${prepare_boot_cache}" | sed "s/^/$submenu_indentation/"
   fi
-  message="$(gettext_printf "Loading Linux %s ..." ${version})"
+
+  if [ "x${GRUB_ROOT_FS_RO}" = "xtrue" ]; then
+    grub_root_fs_mode=ro
+  else
+    grub_root_fs_mode=rw
+  fi
+
+  message="$(gettext_printf "Loading Linux %s ..." "${version}")"
   sed "s/^/$submenu_indentation/" << EOF
 	echo	'$(echo "$message" | grub_quote)'
-	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} rw ${args}
+	linux	${rel_dirname}/${basename} root=${linux_root_device_thisversion} ${grub_root_fs_mode} ${args}
 EOF
   if test -n "${initrd}" ; then
     # TRANSLATORS: ramdisk isn't identifier. Should be translated.
@@ -192,6 +192,19 @@
 	done ;;
 esac
 
+case "x$machine" in
+    aarch64)
+	list=
+	for i in /boot/vmlinuz-* /vmlinuz-* /boot/kernel-* /boot/Image ; do
+	    if grub_file_is_not_garbage "$i" ; then list="$list $i" ; fi
+	done ;;
+    *) 
+	list=
+	for i in /boot/vmlinuz-* /boot/vmlinux-* /vmlinuz-* /vmlinux-* /boot/kernel-* /boot/Image ; do
+                  if grub_file_is_not_garbage "$i" ; then list="$list $i" ; fi
+	done ;;
+esac
+
 case "$machine" in
     i?86) GENKERNEL_ARCH="x86" ;;
     mips|mips64) GENKERNEL_ARCH="mips" ;;
@@ -202,6 +211,29 @@ while [ "x$list" != "x" ] ; do
   alt_version=`echo $version | sed -e "s,\.old$,,g"`
   linux_root_device_thisversion="${LINUX_ROOT_DEVICE}"
 
+  if test -e "/etc/manjaro-release" ; then
+    kbasename="`echo "${basename}" | sed -e 's,vmlinuz-,linux,g'`"
+    if test -e "${dirname}/${kbasename}.kver" ; then
+ 	version="$(cat ${dirname}/${kbasename}.kver)"
+    fi
+    kbasename2="`echo "${basename}" | sed -e 's,vmlinuz-,linux-,g'`"
+    if test -e "${dirname}/${kbasename2}.kver" ; then
+ 	version="$(cat ${dirname}/${kbasename2}.kver)"
+    fi
+    kbasename3="`echo "${basename}" | sed -e 's,vmlinuz-,linux,g' | sed -e 's,\.,,g'`"
+    if test -e "${dirname}/${kbasename3}.kver" ; then
+ 	version="$(cat ${dirname}/${kbasename3}.kver)"
+    fi
+    kbasename4="`echo "${basename}" | sed -e 's,vmlinuz-,linux-,g' | sed -e 's,\.,,g'`"
+    if test -e "${dirname}/${kbasename4}.kver" ; then
+ 	version="$(cat ${dirname}/${kbasename4}.kver)"
+    fi
+    kbasename5="`echo "${basename}" | sed -e 's,vmlinuz-,,g'`"
+    if test -e "${dirname}/${kbasename5}.kver" ; then
+ 	version="$(cat ${dirname}/${kbasename5}.kver)"
+    fi
+  fi
+
   initrd_early=
   for i in ${GRUB_EARLY_INITRD_LINUX_STOCK} \
 	   ${GRUB_EARLY_INITRD_LINUX_CUSTOM}; do
@@ -211,6 +239,7 @@ while [ "x$list" != "x" ] ; do
   done
 
   initrd_real=
+  initramfs_manjaro="`echo "${basename}" | sed -e 's,vmlinuz,initramfs,g'`"
   for i in "initrd.img-${version}" "initrd-${version}.img" \
 	   "initrd-${alt_version}.img.old" "initrd-${version}.gz" \
 	   "initrd-${alt_version}.gz.old" "initrd-${version}" \
@@ -218,7 +247,9 @@ while [ "x$list" != "x" ] ; do
 	   "initramfs-genkernel-${version}" \
 	   "initramfs-genkernel-${alt_version}" \
 	   "initramfs-genkernel-${GENKERNEL_ARCH}-${version}" \
-	   "initramfs-genkernel-${GENKERNEL_ARCH}-${alt_version}"; do
+	   "initramfs-genkernel-${GENKERNEL_ARCH}-${alt_version}" \
+	   "${initramfs_manjaro}.img" \
+	   "initramfs-linux.img" ; do
     if test -e "${dirname}/${i}" ; then
       initrd_real="${i}"
       break
@@ -285,6 +315,17 @@ while [ "x$list" != "x" ] ; do
 
   linux_entry "${OS}" "${version}" advanced \
               "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
+
+  for i in "${initramfs_manjaro}-fallback.img" "initramfs-${version}-fallback.img" ; do
+    if test -e "${dirname}/${i}" ; then
+      initrd="${i}"
+      gettext_printf "Found initrd fallback image: %s\n" "${dirname}/${initrd}" >&2
+      linux_entry "${OS}" "${version}" fallback \
+                  "${GRUB_CMDLINE_LINUX} ${GRUB_CMDLINE_LINUX_DEFAULT}"
+      break
+    fi
+  done
+
   if [ "x${GRUB_DISABLE_RECOVERY}" != "xtrue" ]; then
     linux_entry "${OS}" "${version}" recovery \
                 "single ${GRUB_CMDLINE_LINUX}"



